runtime/memprof.c

```diff
@@ -206,1126 +1258,921 @@ void caml_memprof_update_suspended(bool s) {
-     ([delete_idx <= len]). */
-  uintnat delete_idx;
-};
-
-#define MIN_ENTRIES_LOCAL_ALLOC_LEN 16
-#define MIN_ENTRIES_GLOBAL_ALLOC_LEN 128
-
-/* Entries for other blocks. This variable is shared across threads. */
-static struct entry_array entries_global =
-  { NULL, MIN_ENTRIES_GLOBAL_ALLOC_LEN, 0, 0, 0, 0 };
+/* An entry_action to scan the user_data root */
 
-/* There are no pending callbacks in [entries_global] before this
-   position ([callback_idx <= entries_global.len]). */
-static uintnat callback_idx;
+static bool entry_scan(entry_t entry, void *data)
+{
+  struct scan_closure *closure = data;
+  closure->f(closure->fdata, entry->user_data, &entry->user_data);
+  if (closure->weak && (entry->block != Val_unit)) {
+    closure->f(closure->fdata, entry->block, &entry->block);
+  }
+  return false;
+}
 
-#define CB_IDLE -1
-#define CB_LOCAL -2
-#define CB_STOPPED -3
+/* An entries_action to scan the config root */
 
-/* Structure for thread-local variables. */
-struct caml_memprof_th_ctx {
-  /* [suspended] is used for masking memprof callbacks when
-     a callback is running or when an uncaught exception handler is
-     called. */
-  int suspended;
-
-  /* [callback_status] contains:
-     - CB_STOPPED if the current thread is running a callback, but
-       sampling has been stopped using [caml_memprof_stop];
-     - The index of the corresponding entry in the [entries_global]
-       array if the current thread is currently running a promotion or
-       a deallocation callback;
-     - CB_LOCAL if the current thread is currently running an
-       allocation callback;
-     - CB_IDLE if the current thread is not running any callback.
-  */
-  intnat callback_status;
-
-  /* Entries for blocks whose alloc callback has not yet been called. */
-  struct entry_array entries;
-} caml_memprof_main_ctx =
-  { 0, CB_IDLE, { NULL, MIN_ENTRIES_LOCAL_ALLOC_LEN, 0, 0, 0, 0 } };
-static struct caml_memprof_th_ctx* local = &caml_memprof_main_ctx;
-
-/* Pointer to the word following the next sample in the minor
-   heap. Equals [Caml_state->young_alloc_start] if no sampling is planned in
-   the current minor heap.
-   Invariant: [caml_memprof_young_trigger <= Caml_state->young_ptr].
- */
-value* caml_memprof_young_trigger;
+static void entries_finish_scan(entries_t es, bool young, void *data)
+{
+  struct scan_closure *closure = data;
+  (void)young;
+  closure->f(closure->fdata, es->config, &es->config);
+}
 
-/* Whether memprof has been initialized.  */
-static int init = 0;
+/* Function called by either major or minor GC to scan all the memprof roots */
 
-/* Whether memprof is started. */
-static int started = 0;
+void caml_memprof_scan_roots(scanning_action f,
+                             scanning_action_flags fflags,
+                             void* fdata,
+                             caml_domain_state *state,
+                             _Bool weak,
+                             _Bool global)
```

Comment:
We use `_Bool` in the header file function prototype so that you don't have to include `stdbool.h` to include the header file (see #11990 and #12196). But I agree that here in the implementation we should feel free to use `bool` (and `true` and `false`).